runtime/common/SimulationState.h

```diff
@@ -0,0 +1,201 @@
+
+/// @brief The `SimulationState` interface provides and extension point
+/// for concrete circuit simulation sub-types to describe their
+/// underlying quantum state data in an efficient manner. The `SimulationState`
+/// provides a handle on the data for clients, with data remaining on GPU
+/// device or CPU memory. The primary goal of this type and its sub-types
+/// is to minimize data transfers for the state.
+class SimulationState {
+protected:
+  /// @brief Utility method to inspect the state_data variant
+  /// and extract the data pointer and size.
+  template <typename T, typename ScalarType = double>
+  auto getSizeAndPtrFromVec(const state_data &data) {
+    if constexpr (!std::is_same_v<T, ScalarType>)
+      throw std::runtime_error("[sim-state] invalid data precision.");
+
+    auto &vec = std::get<std::vector<std::complex<T>>>(data);
+    return std::make_tuple(
+        vec.size(),
+        reinterpret_cast<void *>(const_cast<std::complex<T> *>(vec.data())));
+  };
+
+  /// @brief Utility method to inspect the state_data variant
+  /// and extract the data pointer and size.
+  template <typename T, typename ScalarType = double>
+  auto getSizeAndPtrFromPair(const state_data &data) {
+    if constexpr (!std::is_same_v<T, ScalarType>)
+      throw std::runtime_error("[sim-state] invalid data precision.");
+
+    auto &pair = std::get<std::pair<std::complex<T> *, std::size_t>>(data);
+    return std::make_tuple(pair.second, reinterpret_cast<void *>(pair.first));
+  };
+
+  /// @brief Utility method to inspect the state_data variant
+  /// and extract the data pointer and size.
+  template <typename ScalarType = double>
+  auto getSizeAndPtr(const state_data &data) {
+    auto type = data.index();
+    std::tuple<std::size_t, void *> sizeAndPtr;
+    if (type == 0)
+      sizeAndPtr = getSizeAndPtrFromVec<double, ScalarType>(data);
+    else if (type == 1)
+      sizeAndPtr = getSizeAndPtrFromVec<float, ScalarType>(data);
+    else if (type == 2)
+      sizeAndPtr = getSizeAndPtrFromPair<double, ScalarType>(data);
+    else if (type == 3)
+      sizeAndPtr = getSizeAndPtrFromPair<float, ScalarType>(data);
+    else
+      throw std::runtime_error("unsupported data type for state.");
+
+    return sizeAndPtr;
+  }
+
+  /// @brief Subclass-specific creator method for
+  /// new SimulationState instances. Create from the size
+  /// and data pointer, which may be on GPU device.
+  virtual std::unique_ptr<SimulationState>
+  createFromSizeAndPtr(std::size_t, void *, std::size_t dataType) {
+    throw std::runtime_error(
+        "SimulationState::createFromSizeAndPtr not implemented.");
+  }
+
+public:
+  /// @brief Runtime-known precision for the simulation data
+  enum class precision { fp32, fp64 };
+
+  /// @brief Simulation states are made up of a
+  /// vector of Tensors. Each tensor tracks its data pointer
+  /// and the tensor extents.
+  struct Tensor {
+    void *data = nullptr;
```

Comment:
Should we have a `toHost` method so that users can retrieve the full tensor (e.g., MPS tensor) as host data at once?
(just for convenience since users can also do `cudaMemCpy` in user code)